staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go

```diff
@@ -17,92 +17,29 @@ limitations under the License.
+type Validator interface {
+	Validate(a admission.Attributes, o admission.ObjectInterfaces, params runtime.Object) ([]policyDecision, error)
 }
 
-// PolicyDefinition is an interface for internal policy binding type.
-// Implemented by mock/testing types, and to be implemented by the public API
-// types once they have completed API review.
-//
-// The interface closely mirrors the format and functionality of the
-// PolicyDefinition proposed in the KEP.
-type PolicyDefinition interface {
-	runtime.Object
-
+// ValidatorCompiler is Dependency Injected into the PolicyDefinition's `Compile`
+// function to assist with converting types and values to/from CEL-typed values.
+type ValidatorCompiler interface {
 	// Matches says whether this policy definition matches the provided admission
 	// resource request
-	Matches(a admission.Attributes) bool
+	DefinitionMatches(a admission.Attributes, o admission.ObjectInterfaces, definition *v1alpha1.ValidatingAdmissionPolicy) (bool, error)
 
-	Compile(
-		// Definition is provided with a converter which may be used by the
-		// return evaluator function to convert objects into CEL-typed objects
-		objectConverter ObjectConverter,
-		// Injected RESTMapper to assist with compilation
-		mapper meta.RESTMapper,
-	) (EvaluatorFunc, error)
-
-	// GetParamSource returns the GVK for the CRD used as the source of
-	// parameters used in the evaluation of instances of this policy
-	// May return nil if there is no paramsource for this definition.
-	GetParamSource() *schema.GroupVersionKind
-
-	// GetFailurePolicy returns how an object should be treated during an
-	// admission when there is a configuration error preventing CEL evaluation
-	GetFailurePolicy() FailurePolicy
-}
-
-// PolicyBinding is an interface for internal policy binding type. Implemented
-// by mock/testing types, and to be implemented by the public API types once
-// they have completed API review.
-//
-// The interface closely mirrors the format and functionality of the
-// PolicyBinding proposed in the KEP.
-type PolicyBinding interface {
-	runtime.Object
-
-	// Matches says whether this policy binding matches the provided admission
+	// Matches says whether this policy definition matches the provided admission
 	// resource request
-	Matches(a admission.Attributes) bool
+	BindingMatches(a admission.Attributes, o admission.ObjectInterfaces, definition *v1alpha1.ValidatingAdmissionPolicyBinding) (bool, error)
 
-	// GetTargetDefinition returns the Namespace/Name of Policy Definition used
-	// by this binding.
-	GetTargetDefinition() (namespace, name string)
-
-	// GetTargetParams returns the Namespace/Name of instance of TargetDefinition's
-	// ParamSource to be provided to the CEL expressions of the definition during
-	// evaluation.
-	// If TargetDefinition has nil ParamSource, this is ignored.
-	GetTargetParams() (namespace, name string)
+	// Compile is used for the cel expression compilation
+	Compile(
```

Comment:
The compilation err embedded in the validator currently. Since we wanna allow the rest of expressions who passed compilation continue with validation. So later, we could handle compilation err and validation err together

Reply:
Maybe another refactor on compilation side to compile everything at once and return compilation err back before validate? cc @jpbetz